torch/distributed/fsdp/_flat_param.py

```diff
@@ -0,0 +1,474 @@
+        self._param_infos = tuple(param_infos)
+        self._numels = tuple(numels)
+        self._shapes = tuple(shapes)
+        self._prefixed_param_names = tuple(prefixed_param_names)
+        self._shared_param_infos = tuple(shared_param_infos)
+        self._is_sharded = False
+        self._unsharded_size = self.size()
+
+    @property
+    def _flat_param_name(self) -> str:
+        """
+        Returns a name for the flattened parameter that uniquely identifies
+        it in a module hierarchy wrapped by a top-level FSDP instance. The name
+        is constructed from the fully-prefixed names of the original parameters
+        comprising the flattened parameter with "." replaced with "_" since
+        parameter names cannot contain ".".
+        """
+        return ",".join(self._prefixed_param_names).replace(".", "_")
+
+
+class FlatParamHandle:
+    """
+    This handle manages a flattened parameter (:class:`FlatParameter`).
+
+    Args:
+        params (Sequence[nn.Parameter]): The parameters to use for the
+            flattened parameter.
+        module (nn.Module): A module that is the root of the subtree containing
+            all parameters in ``params``; for non-recursive wrapping, this must
+            be the top-level module, while for recursive wrapping, this may not
+            necessarily be the top-level module.
+    """
+    def __init__(
+        self,
+        params: Sequence[nn.Parameter],
+        module: nn.Module,
+    ) -> None:
+        super().__init__()
+        self._init_flat_param(module, params)
+        self._unflatten(as_params=False)
+
+    def _init_flat_param(
+        self,
+        module: nn.Module,
+        params: Sequence[Optional[nn.Parameter]],
+    ) -> None:
+        """
+        Initializes the flattened parameter ``self.flat_param`` by flattening
+        the parameters in ``params`` into a single :class:`FlatParameter` and
+        saves relevant metadata. Shared parameters are only included in the
+        flattened parameter once.
+
+        This checks that all comprising parameters have the same dtype and
+        ``requires_grad`` and does not support nested construction of
+        :class:`FlatParameter` s.
+
+        Args:
+            See the Args in the class docstring.
+        """
+        params_set = set(params)
+        params_set.discard(None)
+        assert len(params_set) > 0
+        param_infos: List[ParamInfo] = []
+        numels: List[int] = []
+        shapes: List[torch.Size] = []
+        prefixed_param_names: List[str] = []
+        shared_param_infos: List[SharedParamInfo] = []
+        shared_param_memo: Dict[nn.Parameter, Tuple[nn.Module, str, str]] = {}
+        params_to_flatten: List[nn.Parameter] = []
+        dtype: Optional[torch.dtype] = None
+        requires_grad: Optional[bool] = None
+        for submodule_name, submodule in module.named_modules():
+            for param_name, param in submodule.named_parameters(recurse=False):
```

Comment:
I think we have to loop starting from the root module because we need the fully prefixed name starting from the root for checkpointing.

The only other option I can think of is caching the prefixed module names across separate `FlatParameter` constructions, but if there exists just one module needed to construct the `FlatParameter` that is not in the cached mapping, then we need to do the full module walk from the root in the worst case.

I think this iteration should not be an issue since even for large models, the number of modules is relatively small (maybe in the order of hundreds at most?), which can be iterated over pretty quickly even in Python.